### PR TITLE
feat: add app.getApplicationNameForProtocol API

### PR DESCRIPTION
**Pull request by ajmacd**
#### Description of Change
This API returns the name of the application registered as the default handler for a protocol (e.g. `irc://`). We're going to use this functionality in Slack to enable a "native app linking" feature, whereby Slack apps will optionally provide URLs to native applications alongside http URLs, which we'll attempt to launch if the correct native application is installed on the user's machine.

It fits well with the other protocol functionality provided by the `Browser` class. `GetApplicationNameForUrl` is an existing method in Chromium:
https://chromium.googlesource.com/chromium/src/+/ed716f08f330dda84f38b7f40f3e381cd6af571a/chrome/browser/shell_integration.h#65

and my initial attempt was simply to call down to it. Sadly, this would appear to require depending on the entire `//chrome/browser` module, which brings in something like a highly undesirable 3k files. Instead I copied the implementations out to Electron, which again is consistent with some of the other methods having similar or identical implementations in Chromium:
* `SetAsDefaultProtocolClient`
* `IsDefaultProtocolClient`

The Chromium Linux implementation is fairly useless, simply returning `"xdg-open"`:
https://chromium.googlesource.com/chromium/src/+/eaa08411bd1f60a4e9d4b3253ba4710ae05bff3c/chrome/browser/shell_integration_linux.cc#720

but I discovered a useful implementation relying on `xdg-mime` wasn't all that hard. I'd like to upstream this to Chromium later. (Again, we won't be relying on that implementation, but just to share the spoils as it were.) This also avoids some `DCHECK`s I hit around disallowed blocking calls on Linux, in both the new and existing functionality. 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

notes: Added `app.getApplicationNameForProtocol()` API

### CODE REVIEW

File: shell/browser/browser_linux.cc [hunk 1]
```diff
@@ -113,6 +131,16 @@ bool Browser::RemoveAsDefaultProtocolClient(const std::string& protocol,
   return false;
 }
 
+base::string16 Browser::GetApplicationNameForProtocol(const GURL& url) {
+  std::vector<std::string> argv;
```
Comment (codebytere) [inline shell/browser/browser_linux.cc]: kinda-nit: for simplicity we can probably just do:
```
const std::string& handler = std::string("x-scheme-handler/") + url.scheme();
argv = {"xdg-mime", "query", "default", handler};
```

File: spec-main/api-app-spec.ts [hunk 2]
```diff
@@ -876,6 +876,32 @@ describe('app module', () => {
         })
       })
     })
+
+    it('sets the default client such that getApplicationNameForProtocol returns Electron', (done) => {
+      app.setAsDefaultProtocolClient(protocol);
+      expect(app.getApplicationNameForProtocol(`${protocol}://`)).to.equal('Electron')
+      done()
```
Comment (codebytere) [inline spec-main/api-app-spec.ts]: if this isn't an explicitly async new method call (docs don't indicate it is unless i missed something) we probably don't need this `done` here

File: spec-main/api-app-spec.ts [hunk 3]
```diff
@@ -876,6 +876,32 @@ describe('app module', () => {
...
+      const protocols = [
+        'http://',
+        'https://',
+      ]
+      // We can't expect particular app names here, but these protocols should
+      // at least have _something_ registered.
+      protocols.forEach((protocol) => {
+        expect(app.getApplicationNameForProtocol(protocol)).to.not.equal('')
+      })
+      done()
```
Comment (codebytere) [inline spec-main/api-app-spec.ts]: same as above

File: docs/api/app.md [hunk 4]
```diff
@@ -770,6 +770,21 @@ macOS machine. Please refer to
 
 The API uses the Windows Registry and LSCopyDefaultHandlerForURLScheme internally.
 
+### `app.getApplicationNameForProtocol(url)`
+
+* `url` String - a URL with the protocol name to check. Unlike the other
+  methods in this family, this accepts an entire URL, including `://` at a
+  minimum (e.g. `https://`).
+
+Returns String - Name of the application handling the protocol, or an empty
```
Comment (codebytere) [inline docs/api/app.md]:
```suggestion
Returns `String` - Name of the application handling the protocol, or an empty
```

File: spec-main/api-app-spec.ts (hunk 2 above)
Comment (codebytere) [inline spec-main/api-app-spec.ts]: if this isn't an explicitly async new method call (docs don't indicate it is unless i missed something?) we probably don't need this `done` here

File: spec-main/api-app-spec.ts (hunk 3 above)
Comment (codebytere) [inline spec-main/api-app-spec.ts]: same as above

File: shell/browser/browser_linux.cc (hunk 1 above)
Comment (codebytere) [inline shell/browser/browser_linux.cc]: kinda-nit: for simplicity we can probably just do:
```cpp
const std::string& handler = std::string("x-scheme-handler/") + url.scheme();
argv = {"xdg-mime", "query", "default", handler};
```

File: shell/browser/browser_mac.mm [hunk 5]
```diff
@@ -131,6 +132,22 @@
...
 }
 
+base::string16 Browser::GetApplicationNameForProtocol(const GURL& url) {
+  NSURL* ns_url = [NSURL
+      URLWithString:base::SysUTF8ToNSString(url.possibly_invalid_spec())];
+  base::ScopedCFTypeRef<CFErrorRef> out_err;
+  base::ScopedCFTypeRef<CFURLRef> openingApp(LSCopyDefaultApplicationURLForURL(
+      (CFURLRef)ns_url, kLSRolesAll, out_err.InitializeInto()));
+  if (out_err) {
+    // likely kLSApplicationNotFoundErr
```
Comment (codebytere) [inline shell/browser/browser_mac.mm]: let's `LOG(ERROR)` this error here?

**Comment by miniak** [comment]
I am not sure if this API belongs to the app module, it does not operate on the current application instance.

File: spec-main/api-app-spec.ts (hunk 2 above)
Comment (ajmacd) [reply spec-main/api-app-spec.ts]: Ah right, thanks.

File: shell/browser/browser_linux.cc (hunk 1 above)
Comment (ajmacd) [reply shell/browser/browser_linux.cc]: Good idea. Updated the other similar uses as well.

**Comment by ajmacd** [comment]
> I am not sure if this API belongs to the app module, it does not operate on the current application instance. What about moving it to systemPreferences instead?

I can see your point, and in a vacuum, I agree, it seems like this new API would be a better fit for `systemPreferences`. However, it's closely related to the existing protocol APIs and I think they should be grouped together. Thoughts?

**Comment by ajmacd** [comment]
Hello fine reviewers: any more thoughts on my comments? Happy to discuss further if you disagree!

**Comment by ajmacd** [comment]
Sorry should have @codebytere and @miniak in the last comment ^ 🙇

**Comment by ajmacd** [comment]
@codebytere, @miniak any comments on this? I'd like to land it, but am happy to discuss further changes.

**Comment by ajmacd** [comment]
@zcbenz would you be OK with landing this, or would you prefer to wait for @codebytere and @miniak's approval?

**Comment by zcbenz** [comment]
If no opposition I would like to land this.

@ajmacd Do you mind rebasing this branch on master as there is conflict?

**Comment by ajmacd** [comment]
I actually chatted quickly with @codebytere about this today. She mentioned wanting to bring it up at the API working group meeting.

**Comment by codebytere** [comment]
@ajmacd this needs a final conflict fix: 

```
Unhandled exception in main spec runner: electron/spec-main/api-app-spec.ts(866,43): error TS2304: Cannot find name 'isCI'.
```

**Comment by ajmacd** [comment]
@codebytere I think we're good to go here. The build-linux failure is unrelated to this change.

**Comment by jkleinsc** [comment]
Approved by the Releases WG for backport to 8-x-y

**Pull request merged**